lib/eex/lib/eex/tokenizer.ex

```diff
@@ -123,24 +125,28 @@ defmodule EEx.Tokenizer do
   defp token_name('>-' ++ rest) do
     rest = Enum.reverse(rest)
 
-    # Tokenize the remaining passing check_terminators as
-    # false, which relax the tokenizer to not error on
-    # unmatched pairs. Then, we check if there is a "fn"
-    # token and, if so, it is not followed by an "end"
-    # token. If this is the case, we are on a start expr.
-    case :elixir_tokenizer.tokenize(rest, 1, file: "eex", check_terminators: false) do
-      {:ok, tokens} ->
-        tokens = Enum.reverse(tokens)
-        fn_index = fn_index(tokens)
-
-        if fn_index && end_index(tokens) > fn_index do
-          :start_expr
-        else
+    if starts_with_end_expr?(rest) do
+      :middle_expr
+    else
+      # Tokenize the remaining passing check_terminators as
+      # false, which relax the tokenizer to not error on
+      # unmatched pairs. Then, we check if there is a "fn"
+      # token and, if so, it is not followed by an "end"
+      # token. If this is the case, we are on a start expr.
+      case :elixir_tokenizer.tokenize(rest, 1, file: "eex", check_terminators: false) do
+        {:ok, tokens} ->
```

Comment:
Hi @LostKobrakai!

I have cleaned up your commit a bit and then I noticed that your implementation actually has a bug. It will match if the code starts with a variable named `endgame`, for example.

I think that instead of looking at the buffer, you should look if the first token is the end token. It will be the same result, except you will rely on the Elixir tokenizer to handle it properly.

Reply:
Great, I'll take a look at it. I wasn't really sure how the elixir_tokenizer behaves, so I tried to stay clear of it :D

Reply:
I've gone and replace the check with looking at the results of the tokenizer and added the same improvements to tokens ending in `do %>`